### DISCUSSION TITLE
How to export all Org files in a directory + How to generate offline navigable HTML files using Hugo?

### DISCUSSION THREAD

**Discussion by buhtz** [kaushalmodi/ox-hugo#585] [Q&A] [answered]
From the documentation it is a bit unclear for me how I can export all existing org files I have created with org-roam-v2.
I know about `org-hugo-export-wim-to-md` which export the current buffer only.

I want to export all org files without opening one in a buffer.

Maybe the term "subtree" has something to do with this? But I do not understand what this is.

My goal is not a blog or nice looking website. I simply want to read all my org-roam-v2 created nodes as HTML files.

**Comment by kaushalmodi** [comment]
> Maybe the term "subtree" has something to do with this? 

No, a "subtree" is an Org mode subtree. For example, see subtrees A and B below.

```org
* A
Something
* B
Something
```

**Reply by buhtz** [reply]
I would call this a head or sub heading. Am I right?

**Reply by kaushalmodi** [reply]
- Just "A" is the heading.
- But everything A and under that is a "subtree".

The "subtree" term is used consistently in Org manual. See the **(org) Headlines** Info node.

It is also the "subtree" that you export when you use the default `C-c C-e C-s ..` binding in Org to export using any exporter:

![image](https://user-images.githubusercontent.com/3578197/157756857-c412d094-c0ae-40c4-814d-c49555985f45.png)

**Comment by kaushalmodi** [comment]
> My goal is not a blog or nice looking website. 

Yeah, that's unrelated to `ox-hugo`. The "looks" of your site depend solely on your Hugo layout files or the theme.

**Comment by kaushalmodi** [comment]
> I want to export all org files without opening one in a buffer.

This is out of the scope of `ox-hugo`. Typically you would call emacs in batch mode, loop through all the Org files you want to export and call the "wim" export function on each. 

Here's one [example](https://fosstodon.org/@shom/107879184947755795) of how to export all files from the command line (Credit: @shombando).

**Reply by buhtz** [reply]
The example looks easy but I do not see how it open all org files? Are you sure that this example shows how to export **all** files?

I simply modified this line
`(with-current-buffer (find-file "./content.org")`
to this
`(with-current-buffer (find-file "~/test_org")`

But of course it treats `~/test_org` as the org file to open but it is the directory where all org files are stored in.

Shouldn't there be a loop or something like that? I have no idea about Lisp I am more a Python type. ;)

**Reply by kaushalmodi** [reply]
Oops, you are right, That code does not do the looping, but it's a good example of how to call the exporter from the command line. Now you only need to loop through all Org files 😄

**Reply by kaushalmodi** [reply]
@Codeberg-AsGithubAlternative-buhtz Here's a example that loops through all the Org files and exports them (credit: @bphenriques)

- [Makefile target](https://github.com/bphenriques/knowledge-base/blob/e4c9e334ad605c91e0c813ec9f9e95c36e7fa8a3/Makefile#L22)
- [`build/export-all` elisp function](https://github.com/bphenriques/knowledge-base/blob/e4c9e334ad605c91e0c813ec9f9e95c36e7fa8a3/tools/init.el#L65-L73)

**Reply by buhtz** [reply]
I tried to contact the author via mastodon/fosstodon but I am not able to get my account their working.
I also tried some "walk throw directories" and "loop over files" examples in Lisp. But I miss Lisp basic knowledge. I do not get it. ;)

Is this use case so unusual that you do not implement this in ox-hugo directly? Something like `export-my-org-roam-files` using the current setup `org-roam-directory`?

**Reply by kaushalmodi** [reply]
> Is this use case so unusual that you do not implement this in ox-hugo directly?

`ox-hugo` exports a file or a buffer. The act of exporting typically begins from an Org buffer.

Batch exporting is more related to command-line machinery. People may be bash scripts, Makefiles, Python scripts, etc. to simply call emacs in batch mode and run `org-hugo-export-wim-to-md`. I hope my above comment helps you achieve this: https://github.com/kaushalmodi/ox-hugo/discussions/585#discussioncomment-2334787.

> Something like `export-my-org-roam-files` using the current setup `org-roam-directory`?

The export flow has little to do with org-roam. I cannot bake in a solution specific to that. If you or someone decides to write a blog post on how they set up exporting all org-roam files using `ox-hugo`, I can link that in the ox-hugo manual somewhere.

---

Coming up with a robust way to do that for users coming from various OS, various use cases (org-roam, etc) and then maintaining that is really out of the scope of this project.

**Reply by buhtz** [reply]
Thank you so much. I do not know about `knowledge-base-dir` so I modified line nr 67 (without really knowing what I am doing) to this
`(let (search-path org-roam-directory)`
But then I get `Wrong type argument: stringp, nil`

And I do not understand what the `"org/"` is there for.

**Reply by bphenriques** [reply]
If ` org-roam-directory` is the path that contains your mix of subpaths and `.org` files, you can do this:
```elisp
(defun build/export-all ()
  (dolist (org-file (directory-files-recursively org-roam-directory "\.org$"))
     (with-current-buffer (find-file org-file)
			   (message (format "[build] Exporting %s" org-file))
			   (org-hugo-export-wim-to-md :all-subtrees nil nil nil))))
```

Removed the non-essential bits.

Edit: you actually need to update  `org-ird-extra-files` as well

**Reply by bphenriques** [reply]
In my use-case, the script is the CI/CD pipeline therefore needs to be self-contained so it has extra stuff.

It depends on the `knowledge-base-dir` which is the root of my org-roam [repo](https://github.com/bphenriques/knowledge-base/) and is set using a environment variable (Lines 22-27). Then:
- Set ` org-hugo-base-dir` to `knowledge-base-dir` to avoid setting up per file - Line 57
- Ensure that all `org` files are registered so that org-roam may find them: `(setq  org-id-extra-files (directory-files-recursively knowledge-base-dir "\.org$")` - Line 59
- Using ox-hugo, export all files under  `<knowledge-base-dir>/org` - The `build/export-all`. Note that it cant be just `<knowledge-base-dir>`in my case because I do not want to export the `README.org` file.

If you do not need this to be in a CI/CD pipeline, the sample above may suffice. 

Hope it broadly explains! Try out and experiment, each case is unique :)

**Reply by kaushalmodi** [reply]
Here's another version (not thoroughly tested) derived from @bphenriques 's work:

```emacs-lisp
(defun ox-hugo/export-all (&optional org-files-root-dir dont-recurse)
  "Export all Org files (including nested) under ORG-FILES-ROOT-DIR.

All valid post subtrees in all Org files are exported using
`org-hugo-export-wim-to-md'.

If optional arg ORG-FILES-ROOT-DIR is nil, all Org files in
current buffer's directory are exported.

If optional arg DONT-RECURSE is nil, all Org files in
ORG-FILES-ROOT-DIR in all subdirectories are exported. Else, only
the Org files directly present in the current directory are
exported.  If this function is called interactively with
\\[universal-argument] prefix, DONT-RECURSE is set to non-nil.

Example usage in Emacs Lisp: (ox-hugo/export-all \"~/org\")."
  (interactive)
  (let* ((org-files-root-dir (or org-files-root-dir default-directory))
         (dont-recurse (or dont-recurse (and current-prefix-arg t)))
         (search-path (file-name-as-directory (expand-file-name org-files-root-dir)))
         (org-files (if dont-recurse
                        (directory-files search-path :full "\.org$")
                      (directory-files-recursively search-path "\.org$")))
         (num-files (length org-files))
         (cnt 1))
    (if (= 0 num-files)
        (message (format "No Org files found in %s" search-path))
      (progn
        (message (format (if dont-recurse
                             "[ox-hugo/export-all] Exporting %d files from %S .."
                           "[ox-hugo/export-all] Exporting %d files recursively from %S ..")
                         num-files search-path))
        (dolist (org-file org-files)
          (with-current-buffer (find-file-noselect org-file)
            (message (format "[ox-hugo/export-all file %d/%d] Exporting %s" cnt num-files org-file))
            (org-hugo-export-wim-to-md :all-subtrees)
            (setq cnt (1+ cnt))))
        (message "Done!")))))
```

**Reply by buhtz** [reply]
> If ` org-roam-directory` is the path that contains your mix of subpaths and `.org` files, you can do this:
> 
> ```emacs-lisp
> (defun build/export-all ()
>   (dolist (org-file (directory-files-recursively org-roam-directory "\.org$"))
>      (with-current-buffer (find-file org-file)
> 			   (message (format "[build] Exporting %s" org-file))
> 			   (org-hugo-export-wim-to-md :all-subtrees nil nil nil))))
> ```
> 
> Removed the non-essential bits.
> 
> Edit: you actually need to update `org-ird-extra-files` as well

Thanks for that code. I did `C-h v org-id-extra-files` but do not understand why and how this is related to my case. I would say there are no "extra files" just everything in the org-roam-directory is important for me.

It is hard for me to understand the debug output. Line numbers or other location indicators are missing or not understandable by me. Which part of the code causing the error?
```
eval: Wrong number of arguments: ((t) nil (let ((--dolist-tail-- (directory-files-recursively org-roam-directory ".\
    org$"))) (while --dolist-tail-- (let ((org-file (car --dolist-tail--))) (save-current-buffer (set-buffer (find-file\
     org-file)) (message (format "[build] Exporting %s" org-file)) (org-hugo-export-wim-to-md :all-subtrees nil nil nil\
    )) (setq --dolist-tail-- (cdr --dolist-tail--)))))), 1
```
Just for completeness I copy & paste my code here. I have done a `diff` to be sure the code is identical with yours.
```emacs-lisp
(require 'ox-hugo)
(defun build/export-all ()
  (dolist (org-file (directory-files-recursively org-roam-directory "\.org$"))
    (with-current-buffer (find-file org-file)
      (message (format "[build] Exporting %s" org-file))
      (org-hugo-export-wim-to-md :all-subtrees nil nil nil))))
```

**Reply by buhtz** [reply]
> Here's another version (not thoroughly tested) derived from @bphenriques 's work:
> 
> ```emacs-lisp
> (defun ox-hugo/export-all (&optional org-files-root-dir dont-recurse)
>   "Export all Org files (including nested) under ORG-FILES-ROOT-DIR.
>   ; ...
> ```

This seems to work. I have md-files with links in them.
I also could setup a quick and dirty hugo-thing and click on that links. It works.

But where is the HTML content? I do not want to run hugo as a server. I need hugo (or something else) to transform the md-files into html-files. Maybe I [misunderstood](https://discourse.gohugo.io/t/where-are-the-html-files/37603?u=codeberg-asgithubalt) the therm "static site generator"?

I got it. The hugo introduction is but to far over the border and says `hugo server`. But just doing `hugo` will generate HTML files in the `public` folder. But in that case the links are dead. Each org-roam note becomes an `index.html` in it's own folder.
Example with only two notes:
```
$ tree content
content
└── posts
    ├── 20220309165918-bar.md
    └── 20220309165928-foo.md
```

This becomes
```
$ tree public
public
├── 404.html
├── ananke
├── index.html
├── index.xml
├── posts
│   ├── 20220309165918-bar
│   │   └── index.html
│   ├── 20220309165928-foo
│   │   └── index.html
│   ├── index.html
│   ├── index.xml
│   └── page
│       └── 1
│           └── index.html
...
```

The url-link to `Bar` in `Foo` (`public/posts/20220309165928-foo/index.html`) points to `public/posts/20220309165918-bar` (a folder). A browser can not handle this because no server is running serving him a `index.html`. I assume the HTML content is useless without a running server. Am I right?

If this is true I would not call hugo a "static site generator".

I am frustrated. I am afraid I have to do all this on my own with a Python script beginning from parsing the org-roam-files. This would give me more control and power over the process. Maybe my use case is to unusual?

**Reply by kaushalmodi** [reply]
Well, `hugo` **is** a static site generator. You'll just need to learn that tool and certain concepts on how HTML pages work. You are probably generating pages with absolute URLs instead of relative URLs. Hugo can generate both kind, but then you need to pick a theme that ensures that the pages can be viewed offline (i.e. uses relative permalinks). 

I would suggest you asking a question on Hugo forum (there are a lot of helpful and friendly folks there. They helped me as I was researching on Hugo and developing ox-hugo. I had no idea about relative and absolute URLs when I started as well): https://discourse.gohugo.io/. Ask if people are aware of Hugo themes that can generate offline HTML content; that you want to generate pages for offline viewing and don't plan to host them on any server. 

> A browser can not handle this because no server is running serving him a index.html

I don't see why an index.html cannot be viewed offline. The only difference that a page being `foo/index.html` makes vs `foo.html` is that the former can be viewed on `file://foo/` url while the latter can be viewed on `file://foo.html`. In Hugo lingo, the urls with ".html" are called "ugly" urls (you can look up uglyURLs in Hugo docs) and the urls ending in "/" are "pretty" URLs. Look up on how to make Hugo generate pages with ugly urls. 

> I am afraid I have to do all this on my own with a Python script beginning from parsing the org-roam-files. This would give me more control and power over the process. 

Certainly! If you go that route, eventually you will understand why those links in public/ did not work because you'll need to learn HTML :)

> Maybe my use case is to unusual?

No, Hugo deploys its [documentation site releases](https://github.com/gohugoio/hugoDocs/releases/tag/v0.94.0) .tar.gz files such that they are viewable offline too.

**Reply by buhtz** [reply]
Thanks for your thoughts and ideas.

**Reply by kaushalmodi** [reply]
@Codeberg-AsGithubAlternative-buhtz I believe [my reply](https://discourse.gohugo.io/t/where-are-the-html-files/37603/7?u=kaushalmodi) should provide you a minimal example of how to design the Hugo config and layout files to make your generated HTML pages locally navigable.

_**Note:** There's a difference between `static` and `offline` HTML pages. A static site generator generates HTML pages that are self-contained and don't require to dynamically fetch page content from external databases (like Wordpress and others do)._

**Reply by buhtz** [reply]
> ```emacs-lisp
> (defun ox-hugo/export-all (&optional org-files-root-dir dont-recurse)
>   "Export all Org files (including nested) under ORG-FILES-ROOT-DIR.
> ...
There is one little problem with that function. All opened buffers with the org files are kept open. Can this code be modified that way that the buffer is closed after its content is exported to md?

**Reply by kaushalmodi** [reply]
> All opened buffers with the org files are kept open. 

That's what I meant earlier that it's not possible to support this officially. 

You'll need to first check if a buffer is already open before it's exported and then close that buffer only if it wasn't open before the export. 

I'll encourage to read up on the buffer management functions from emacs lisp Info manual, ask questions on emacs sub reddit or emacs gnu help and learn more. 

This function is best used if you call it from an emacs session in batch mode, through a shell script or Makefile. Then you don't have to worry about open files.

**Comment by kaushalmodi** [comment, marked as answer]
@Codeberg-AsGithubAlternative-buhtz I am summarizing the solutions to your questions here (for future, please open separate Discussion topics for each question).

- [How to export all Org posts in a directory to Markdown using `ox-hugo`?](https://github.com/kaushalmodi/ox-hugo/discussions/585#discussioncomment-2335203)
- [How to use Hugo so that it generates HTML pages that can be navigated without a server?](https://discourse.gohugo.io/t/where-are-the-html-files/37603/7?u=kaushalmodi)

**Comment by buhtz** [comment]
Please allow me to warm up the topic. As stated in [this comment](https://discourse.gohugo.io/t/where-are-the-html-files/37603/5?u=codeberg-asgithubalt) two years ago I now [released my own org-to-html-converter](https://codeberg.org/buhtz/hyperorg).

**Comment by bphenriques** [comment]
Don't know if already shared here. In my now archived digital garden I setup a automated GHA to publish my org-files, it should be portable: https://github.com/bphenriques/knowledge-base/blob/master/.github/workflows/build-deploy.yml
